opt/scopeserver/dataserver/utils/loom.py

```diff
@@ -43,36 +47,60 @@ def __init__(self, partial_md5_hash, file_path, abs_file_path, loom_connection,
         # Metrics
         self.nUMI = None
         self.species, self.gene_mappings = self.infer_species()
-        self.ss_pickle_name = os.path.join(os.path.dirname(self.abs_file_path), partial_md5_hash + ".ss_pkl")
+        self.ss_pickle_name = self.abs_file_path.with_suffix(".ss_pkl")
+        self.load_ss(self.ss_pickle_name)
 
+    def load_ss(self, ss_pickle_name: Path) -> None:
         try:
-            with open(self.ss_pickle_name, "rb") as fh:
-                logger.debug(f"Loading prebuilt SS for {file_path} from {self.ss_pickle_name}")
+            with open(ss_pickle_name, "rb") as fh:
+                logger.debug(f"Loading prebuilt SS for {self.file_path} from {ss_pickle_name}")
                 self.ss = pickle.load(fh)
-                if self.ss.search_space_version < ss.CURRENT_SS_VERISON:
-                    raise AttributeError(
-                        f"Cached search space is an old version or does not exist, upgrading to {ss.CURRENT_SS_VERISON}"
-                    )
                 self.ss.loom = self
                 self.ss.dfh = dfh.DataFileHandler()
-
-        except (EOFError, FileNotFoundError, AttributeError, TypeError):
-            logger.debug(f"Building Search Space for {file_path}")
-            self.ss = ss.SearchSpace(loom=self).build()
-            self.ss.loom = None  # Remove loom connection to enable pickling
-            self.ss.dfh = None
-            logger.debug(f"Built Search Space for {file_path}")
-            with open(self.ss_pickle_name, "wb") as fh:
-                logger.debug(f"Writing prebuilt SS for {file_path} to {self.ss_pickle_name}")
-                pickle.dump(self.ss, fh)
+                try:
+                    assert self.ss.search_space_version == ss.CURRENT_SS_VERISON
+                except AssertionError:
+                    logger.error(
+                        f"Cached search space version {self.ss.search_space_version} is not {ss.CURRENT_SS_VERISON}. Rebuilding search space..."
+                    )
+                    self.build_ss()
+                except AttributeError:
+                    logger.error(f"Search space has no version key and is likely legacy. Rebuilding search space...")
+                    self.build_ss()
```

Comment:
I'll look into moving them over.

I did debate having this error as a single one, but wanted to be able to report the old search space version, which I can't do in one error without 2 if's (on legacy search spaces (no version) your code will fail on the logger because `self.ss.search_space_version` doesn't exit).

Are you insistent on this being collapsed into 1 error?

Reply:
No not fixed on that at all. I would prefer you didn't use exceptions for flow control though.